File: backend/src/services/bridges/healthMonitorBridge.ts
/**
 * HealthMonitor → EventBus 桥接
 *
 * 将 HealthMonitor 的指标采集结果转换为 PerceptionEvent 注入 EventBus。
 * 作为内部指标采集器感知源 (D1.5)。
 *
 * 工作方式：
 * - 注册为 EventBus 感知源
 * - 定期从 HealthMonitor 获取最新快照
 * - 将 HealthMetrics 转换为 type='metric' 的 PerceptionEvent
 * - 当健康评分低于阈值时，生成 type='internal' 的告警事件
 */

import { logger } from '../../utils/logger';
import type { EventBus, Priority } from '../eventBus';
import type { HealthMonitor, HealthMetrics, HealthScore } from '../ai-ops/healthMonitor';

export interface HealthMonitorBridgeConfig {
  /** 采集间隔（毫秒），默认 60000 (60s) */
  collectIntervalMs: number;
  /** 健康评分低于此阈值时生成 internal 告警事件，默认 60 */
  alertScoreThreshold: number;
  /** 是否启用，默认 true */
  enabled: boolean;
}

const DEFAULT_CONFIG: HealthMonitorBridgeConfig = {
  collectIntervalMs: 60_000,
  alertScoreThreshold: 60,
  enabled: true,
};

export class HealthMonitorBridge {
  private readonly eventBus: EventBus;
  private readonly healthMonitor: HealthMonitor;
  private readonly config: HealthMonitorBridgeConfig;
  private timer: ReturnType<typeof setTimeout> | null = null;
  private running = false;

  constructor(
    eventBus: EventBus,
    healthMonitor: HealthMonitor,
    config?: Partial<HealthMonitorBridgeConfig>,
  ) {
    this.eventBus = eventBus;
    this.healthMonitor = healthMonitor;
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * 启动桥接：注册感知源并开始定期采集
   */
  start(): void {
    if (this.running) return;
    if (!this.config.enabled) {
      logger.debug('[HealthMonitorBridge] Disabled, skipping start');
      return;
    }

    // 注册为 EventBus 感知源 (D1.2)
    this.eventBus.registerSource({
      name: 'health-monitor-bridge',
      eventTypes: ['metric', 'internal'],
      schemaVersion: '1.0.0',
    });

    this.running = true;
    this.scheduleNext();
    logger.info('[HealthMonitorBridge] Started');
  }

  /**
   * 停止桥接
   */
  stop(): void {
    this.running = false;
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    logger.info('[HealthMonitorBridge] Stopped');
  }

  /**
   * 手动触发一次采集并发布到 EventBus
   */
  async publishOnce(deviceId?: string): Promise<void> {
    try {
      const metrics = await this.healthMonitor.collectMetrics();
      const score = this.healthMonitor.calculateScore(metrics);
      await this.publishMetricEvent(metrics, score, deviceId);

      // 健康评分低于阈值时，额外发布 internal 告警
      if (score.overall < this.config.alertScoreThreshold) {
        await this.publishHealthAlertEvent(metrics, score, deviceId);
      }
    } catch (error) {
      logger.warn('[HealthMonitorBridge] Failed to publish metrics', { error });
    }
  }

  // ─── 内部方法 ───

  private scheduleNext(): void {
    if (!this.running) return;
    this.timer = setTimeout(async () => {
      await this.publishOnce();
      this.scheduleNext();
    }, this.config.collectIntervalMs);
  }

  private async publishMetricEvent(
    metrics: HealthMetrics,
    score: HealthScore,
    deviceId?: string,
  ): Promise<void> {
    const priority = this.scoreToPriority(score.overall);

    await this.eventBus.publish({
      type: 'metric',
      priority,
      source: 'health-monitor-bridge',
      deviceId,
      payload: {
        cpuUsage: metrics.cpuUsage,
        memoryUsage: metrics.memoryUsage,
        diskUsage: metrics.diskUsage,
        interfaceStatus: metrics.interfaceStatus,
        activeConnections: metrics.activeConnections,
        errorRate: metrics.errorRate,
        avgResponseTime: metrics.avgResponseTime,
        score: score.overall,
        level: score.level,
      },
      schemaVersion: '1.0.0',
    });
  }

  private async publishHealthAlertEvent(
    metrics: HealthMetrics,
    score: HealthScore,
    deviceId?: string,
  ): Promise<void> {
    const priority = score.overall < 30 ? 'critical' as Priority
      : score.overall < 50 ? 'high' as Priority
      : 'medium' as Priority;

    await this.eventBus.publish({
      type: 'internal',
      priority,
      source: 'health-monitor-bridge',
      deviceId,
      payload: {
        alertType: 'low_health_score',
        score: score.overall,
        level: score.level,
        cpuUsage: metrics.cpuUsage,
        memoryUsage: metrics.memoryUsage,
        diskUsage: metrics.diskUsage,
      },
      schemaVersion: '1.0.0',
    });
  }

  private scoreToPriority(score: number): Priority {
    if (score >= 80) return 'info';
    if (score >= 60) return 'low';
    if (score >= 40) return 'medium';
    if (score >= 20) return 'high';
    return 'critical';
  }
}

File: opsevo-python/src/opsevo/services/bridges/health_monitor_bridge.py
"""
HealthMonitorBridge — HealthMonitor → EventBus 桥接

定期从 HealthMonitor 获取指标快照，转换为 PerceptionEvent 注入 EventBus。
健康评分低于阈值时额外生成 internal 告警事件。
"""

from __future__ import annotations

import asyncio
from typing import Any

import structlog

from opsevo.events.event_bus import EventBus
from opsevo.events.types import EventType, PerceptionEvent, Priority

logger = structlog.get_logger(__name__)


class HealthMonitorBridge:
    """HealthMonitor → EventBus 桥接器。"""

    def __init__(
        self,
        event_bus: EventBus,
        health_monitor: Any,
        collect_interval_ms: int = 60_000,
        alert_score_threshold: int = 60,
        enabled: bool = True,
    ) -> None:
        self._event_bus = event_bus
        self._health_monitor = health_monitor
        self._collect_interval = collect_interval_ms / 1000
        self._alert_threshold = alert_score_threshold
        self._enabled = enabled
        self._task: asyncio.Task[None] | None = None
        self._running = False

    def start(self) -> None:
        if self._running:
            return
        if not self._enabled:
            logger.debug("HealthMonitorBridge disabled")
            return
        self._event_bus.register_source(
            "health-monitor-bridge",
            {"event_types": ["metric", "internal"], "schema_version": "1.0.0"},
        )
        self._running = True
        self._task = asyncio.create_task(self._loop())
        logger.info("HealthMonitorBridge started")

    def stop(self) -> None:
        self._running = False
        if self._task and not self._task.done():
            self._task.cancel()
        logger.info("HealthMonitorBridge stopped")

    async def publish_once(self, device_id: str | None = None) -> None:
        try:
            metrics = await self._health_monitor.collect_metrics()
            score = self._health_monitor.calculate_score(metrics)
            await self._publish_metric(metrics, score, device_id)
            overall = score.get("overall", 100) if isinstance(score, dict) else getattr(score, "overall", 100)
            if overall < self._alert_threshold:
                await self._publish_health_alert(metrics, score, device_id)
        except Exception as exc:
            logger.warn("HealthMonitorBridge publish failed", error=str(exc))

    async def _loop(self) -> None:
        while self._running:
            await self.publish_once()
            await asyncio.sleep(self._collect_interval)

    async def _publish_metric(
        self, metrics: Any, score: Any, device_id: str | None
    ) -> None:
        overall = score.get("overall", 100) if isinstance(score, dict) else getattr(score, "overall", 100)
        priority = self._score_to_priority(overall)
        payload: dict[str, Any] = {}
        if isinstance(metrics, dict):
            payload = dict(metrics)
        else:
            for attr in ("cpu_usage", "memory_usage", "disk_usage", "error_rate"):
                if hasattr(metrics, attr):
                    payload[attr] = getattr(metrics, attr)
        payload["score"] = overall
        event = PerceptionEvent(
            type=EventType.METRIC,
            priority=priority,
            source="health-monitor-bridge",
            payload=payload,
            schema_version="1.0.0",
        )
        if device_id:
            event.payload["device_id"] = device_id
        await self._event_bus.publish(event)

    async def _publish_health_alert(
        self, metrics: Any, score: Any, device_id: str | None
    ) -> None:
        overall = score.get("overall", 100) if isinstance(score, dict) else getattr(score, "overall", 100)
        if overall < 30:
            priority = Priority.CRITICAL
        elif overall < 50:
            priority = Priority.HIGH
        else:
            priority = Priority.MEDIUM
        payload: dict[str, Any] = {"alert_type": "low_health_score", "score": overall}
        event = PerceptionEvent(
            type=EventType.INTERNAL,
            priority=priority,
            source="health-monitor-bridge",
            payload=payload,
            schema_version="1.0.0",
        )
        if device_id:
            event.payload["device_id"] = device_id
        await self._event_bus.publish(event)

    @staticmethod
    def _score_to_priority(score: float) -> Priority:
        if score >= 80:
            return Priority.INFO
        if score >= 60:
            return Priority.LOW
        if score >= 40:
            return Priority.MEDIUM
        if score >= 20:
            return Priority.HIGH
        return Priority.CRITICAL
